src/main/scala/is/hail/io/index/IndexWriter.scala

```diff
@@ -0,0 +1,165 @@
+  private val internalEncoder = codecSpec.buildEncoder(InternalNodeBuilder.typ(keyType))(trackedOS)
+
+  private def writeInternalNode(node: InternalNodeBuilder): (Long, Long, Any, Long) = {
+    val fileOffset = trackedOS.bytesWritten
+    rootOffset = fileOffset
+
+    val firstKey = node.firstKey
+    val firstKeyOffset = node.firstOffset
+    val firstIndex = node.firstIndex
+
+    internalEncoder.writeByte(1)
+
+    val regionOffset = node.write(rvb)
+    internalEncoder.writeRegionValue(region, regionOffset)
+    internalEncoder.flush()
+
+    region.clear()
+    node.clear()
+
+    (fileOffset, firstIndex, firstKey, firstKeyOffset)
+  }
+
+  private def writeLeafNode(idx: Long): (Long, Any, Long) = {
+    val fileOffset = trackedOS.bytesWritten
+    val firstKey = leafNode.firstKey
+    val firstOffset = leafNode.firstOffset
+
+    leafEncoder.writeByte(0)
+
+    val regionOffset = leafNode.write(rvb, idx)
+    leafEncoder.writeRegionValue(region, regionOffset)
+    leafEncoder.flush()
+
+    region.clear()
+    leafNode.clear()
+
+    (fileOffset, firstKey, firstOffset)
+  }
+
+  private def write() {
+    def updateInternalNodes(level: Int, fileOffset: Long, firstIndex: Long, firstKey: Any, firstOffset: Long) {
+      if (level == internalNodes.size)
+        internalNodes += new InternalNodeBuilder(keyType)
+
+      val node = internalNodes(level)
+
+      if (node.size == 0) {
+        node.setFirstIndex(firstIndex)
+      }
+
+      node += (fileOffset, firstKey, firstOffset)
+
+      if (node.size == branchingFactor) {
+        val (fileOffset, firstIndex, firstKey, firstKeyOffset) = writeInternalNode(node)
+        updateInternalNodes(level + 1, fileOffset, firstIndex, firstKey, firstKeyOffset)
+      }
+    }
+
+    val idx = elementIdx - leafNode.size
+    val (leafFileOffset, leafFirstKey, leafFirstOffset) = writeLeafNode(idx)
+    updateInternalNodes(0, leafFileOffset, idx, leafFirstKey, leafFirstOffset)
+  }
+
+  private def flush() {
+    val nInternalLevels = math.max(IndexUtils.calcDepth(elementIdx, branchingFactor) - 1, 1) // ensure always one internal node
+
+    def flushInternalNodes(level: Int, fileOffset: Long, firstIndex: Long, firstKey: Any, firstOffset: Long) {
+      if (level != nInternalLevels) {
+        val node = internalNodes(level)
+
+        if (node.size == 0)
+          node.setFirstIndex(firstIndex)
+
+        node += (fileOffset, firstKey, firstOffset)
+
+        assert(node.size > 0 && node.size <= branchingFactor)
+
+        val (nextFileOffset, nextFirstIndex, nextFirstKey, nextFirstKeyOffset) = writeInternalNode(node)
+
+        flushInternalNodes(level + 1, nextFileOffset, nextFirstIndex, nextFirstKey, nextFirstKeyOffset)
+      }
+    }
+
+    if (leafNode.size > 0) {
+      write()
+    }
+    assert(internalNodes.size > 0)
+
+    var level = 0
+    while (level < nInternalLevels) {
+      val node = internalNodes(level)
+      if (node.size > 0) {
```

Comment:
Can this be an assert?  Shouldn't the internal nodes always be non-empty?

Reply:
No, not necessarily. You can have internal nodes close to the leaf node that are empty, but higher level ones still need to be flushed. That's why I look for the first internal node that needs to be flushed.